pkg/domain/infra/abi/images.go

```diff
@@ -748,28 +749,38 @@ func (ir *ImageEngine) Sign(ctx context.Context, names []string, options entitie
 				return err
 			}
 
-			// create signature
-			newSig, err := signature.SignDockerManifest(getManifest, dockerReference.String(), mech, options.SignBy)
-			if err != nil {
-				return errors.Wrapf(err, "error creating new signature")
-			}
-			// create the signstore file
-			signatureDir := fmt.Sprintf("%s@%s=%s", sigStoreDir, manifestDigest.Algorithm(), manifestDigest.Hex())
-			if err := os.MkdirAll(signatureDir, 0751); err != nil {
-				// The directory is allowed to exist
-				if !os.IsExist(err) {
-					logrus.Error(err)
-					return nil
+			if options.All {
+				list, err := manifest.ListFromBlob(getManifest, manifestType)
+				if err != nil {
+					return errors.Wrapf(err, "Error parsing manifest list %q", string(getManifest))
+				}
+				instanceDigests := list.Instances()
+				for _, instanceDigest := range instanceDigests {
+					digest := instanceDigest
+					man, _, err := rawSource.GetManifest(ctx, &digest)
+					if err != nil {
+						return err
+					}
+					newSig, err := createSignature(man, dockerReference.String(), mech, options.SignBy)
+					if err != nil {
+						return err
+					}
+					if err = putSignature(sigStoreDir, newSig, instanceDigest); err != nil {
+						logrus.Errorf("error storing signature for %s, %v", dockerReference.String(), err)
+						return nil
+					}
 				}
-			}
-			sigFilename, err := getSigFilename(signatureDir)
-			if err != nil {
-				logrus.Errorf("error creating sigstore file: %v", err)
 				return nil
 			}
-			err = ioutil.WriteFile(filepath.Join(signatureDir, sigFilename), newSig, 0644)
+
+			// create signature
+			newSig, err := createSignature(getManifest, dockerReference.String(), mech, options.SignBy)
 			if err != nil {
-				logrus.Errorf("error storing signature for %s", rawSource.Reference().DockerReference().String())
+				return err
+			}
+			// create the signstore file
+			if err = putSignature(sigStoreDir, newSig, manifestDigest); err != nil {
+				logrus.Errorf("error storing signature for %s, %v", dockerReference.String(), err)
 				return nil
```

Comment:
(Absolutely non-blocking: This is pre-existing and more or less unrelated, still: I don’t understand why some errors are fatal, and returned, and some are  only logged and otherwise ignored.)

Reply:
If some errors occur before `putSignature`,  return the error since the `putSignature` is not useful in that situation. `putSignature` just logrus the error so it can continue to next image inside the for loop, but this may lead to failing silently.

Reply:
From the point of the user, either the command succeeded in doing everything required, or not. If it doesn’t do everything, the user must notice and just _somehow_; if the command reports that failure in two different ways, I can’t see how that is mores useful than doing it in one way.

And just logging the error makes automation much more difficult — especially if users don’t encounter that case and don’t even _know_ they should be checking for logs.